### PR TITLE
[postgres] reduce unnecessary queries

### LINKED ISSUE #2210: Postgres connector: No need to get CustomTypeMap for API, non-engine purposes
We obtain the postgres connector in various API endpoints. As part of NewPostgresConnector, we load and store the custom types of the source postgres peer. However this custom type map will not be needed for many API endpoints

The SELECT query to get the custom types is an additional operation on source which is unnecessary in these cases. In some busy databases the query can also get locked for a while, resulting in long page load times in the UI

1. Identify places where we get a postgres connector
2. We can put that specific fetch under a flag maybe for the above

### PR DESCRIPTION
**Pull request by heavycrystal**
closes #2210

### CODE REVIEW

File: flow/connectors/postgres/cdc.go [hunk 1]
```diff
@@ -85,21 +85,18 @@ func GetChildToParentRelIDMap(ctx context.Context, conn *pgx.Conn) (map[uint32]u
 		WHERE parent.relkind='p';
 	`
 
-	rows, err := conn.Query(ctx, query, pgx.QueryExecModeSimpleProtocol)
```
Comment (serprex) [inline flow/connectors/postgres/cdc.go:88]: I remember putting this in for a reason

File: flow/shared/postgres.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func GetCustomDataTypes(ctx context.Context, conn *pgx.Conn) (map[uint32]string,
 		var typeID pgtype.Uint32
```
Comment (serprex) [inline flow/shared/postgres.go:66]: Also change this to pgx.ForEachRow

File: flow/connectors/postgres/postgres.go [hunk 3]
```diff
@@ -106,14 +100,25 @@ func NewPostgresConnector(ctx context.Context, env map[string]string, pgConfig *
...
-		customTypesMapping:     customTypeMap,
+		customTypesMapping:     nil,
 		metadataSchema:         metadataSchema,
 		hushWarnOID:            make(map[uint32]struct{}),
 		logger:                 logger,
 		relationMessageMapping: make(model.RelationMessageMapping),
 	}, nil
 }
 
+func (c *PostgresConnector) lazyInitCustomTypesMapping(ctx context.Context) error {
```
Comment (serprex) [inline flow/connectors/postgres/postgres.go:111]: this is really error prone unless this method returns the custom types map & code doesn't directly access

File: flow/connectors/postgres/postgres.go [hunk 4]
```diff
@@ -42,6 +42,7 @@ type PostgresConnector struct {
 	customTypesMapping     map[uint32]string
 	hushWarnOID            map[uint32]struct{}
 	relationMessageMapping model.RelationMessageMapping
+	pgVersion              *shared.PGVersion
```
Comment (serprex) [inline flow/connectors/postgres/postgres.go:45]:
```suggestion
	pgVersion              shared.PGVersion
```
Can treat `0` as uninitialized

File: flow/connectors/postgres/postgres.go (hunk 3 above)
Comment (heavycrystal) [reply flow/connectors/postgres/postgres.go:113]: I'll switch to return map but I'm not sure about the threadsafe requirement, where is that possible? `SetupReplConn` is where I initialize it and that should be before any of the `ReplPing` stuff

Comment (serprex) [reply flow/connectors/postgres/postgres.go:113]: Mostly pointing out that any call to this function needs to have exclusion on connection, but I don't think that's an issue

Comment (heavycrystal) [reply flow/connectors/postgres/postgres.go:113]: can tie it to a mutex if needed in the future, same concern for caching version

Comment (heavycrystal) [reply flow/connectors/postgres/postgres.go:113]: `customTypesMapping` is in some hot paths, keeping it as a field

Comment (serprex) [reply flow/connectors/postgres/postgres.go:113]: it's okay, nil check is cheap

File: flow/connectors/postgres/qrep_query_executor_test.go [hunk 5]
```diff
@@ -214,28 +181,20 @@ func TestAllDataTypes(t *testing.T) {
...
-		t.Fatalf("expected %v, got %v", expectedJSON, record[7].Value())
-	}
+	require.Equal(t, expectedJSON, record[7].Value(), "expected '{\"key\":\"value\"}'")
 
 	actualUUID := record[8].Value().([16]uint8)
-	if !bytes.Equal(actualUUID[:], savedUUID[:]) {
-		t.Fatalf("expected %v, got %v", savedUUID, actualUUID)
-	}
+	require.Equal(t, savedUUID[:], actualUUID[:], "expected savedUUID: %v", savedUUID)
+	// not sure why the values are unequal
```
Comment (serprex) [inline flow/connectors/postgres/qrep_query_executor_test.go:194]: might be caused by PG being microsecond precision. Try rounding `time.Now()` to seconds

**Review by serprex — APPROVED** [review]

**Pull request merged**